### PR TITLE
8267920: Create separate Events buffer for VMOperations

### PR DESCRIPTION
**Pull request by stefank**
The Events classes collect events in a circular buffer that gets dumped into the hs_err files. There are different sections to sort out different types of events. See:
```
  // A log for internal exception related messages, like internal
  // throws and implicit exceptions.
  static ExceptionsEventLog* _exceptions;

  // Deoptization related messages
  static StringEventLog* _deopt_messages;

  // Redefinition related messages
  static StringEventLog* _redefinitions;

  // Class unloading events
  static UnloadingEventLog* _class_unloading;
```
There's also a buffer for non-categorized events:
```
  // A log for generic messages that aren't well categorized.
  static StringEventLog* _messages;
```
I propose that we create a separate buffer for VMOperations. This will make it easier to debug GC related bugs.

With the proposed patch, the hs_err files will now have a section that looks like this.
```
VM Operations (20 events):
Event: 0,186 Executing VM operation: HandshakeAllThreads
Event: 0,186 Executing VM operation: HandshakeAllThreads done
Event: 0,230 Executing VM operation: ZMarkStart
Event: 0,230 Executing VM operation: ZMarkStart done
Event: 0,232 Executing VM operation: HandshakeAllThreads
Event: 0,232 Executing VM operation: HandshakeAllThreads done
Event: 0,232 Executing VM operation: HandshakeAllThreads
Event: 0,232 Executing VM operation: HandshakeAllThreads done
Event: 0,232 Executing VM operation: HandshakeAllThreads
Event: 0,233 Executing VM operation: HandshakeAllThreads done
Event: 0,233 Executing VM operation: ZMarkEnd
Event: 0,233 Executing VM operation: ZMarkEnd done
Event: 0,234 Executing VM operation: HandshakeAllThreads
Event: 0,234 Executing VM operation: HandshakeAllThreads done
Event: 0,234 Executing VM operation: ZVerify
Event: 0,234 Executing VM operation: ZVerify done
Event: 0,234 Executing VM operation: CleanClassLoaderDataMetaspaces
Event: 0,234 Executing VM operation: CleanClassLoaderDataMetaspaces done
Event: 0,235 Executing VM operation: ZRelocateStart
Event: 0,235 Executing VM operation: ZRelocateStart done
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267920](https://bugs.openjdk.java.net/browse/JDK-8267920): Create separate Events buffer for VMOperations

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 668d5fdb16546f7f90ca218657f8089e30b900de
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 1d22bd02b46f1baf573a0f311faf2e4d89cebcb3
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 1d22bd02b46f1baf573a0f311faf2e4d89cebcb3

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4243/head:pull/4243` \
`$ git checkout pull/4243`

Update a local copy of the PR: \
`$ git checkout pull/4243` \
`$ git pull https://git.openjdk.java.net/jdk pull/4243/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4243`

View PR using the GUI difftool: \
`$ git pr show -t 4243`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4243.diff">https://git.openjdk.java.net/jdk/pull/4243.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/events.hpp [hunk 1]
```diff
@@ -414,16 +428,46 @@ inline void EventLogBase<ExtendedStringLogMessage>::print(outputStream* out, Ext
...
+  ~EventMarkWithLogFunction() {
+    if (LogEvents) {
+      log_end();
+    }
+  }
 };
 
+// These end up in the default log.
+typedef EventMarkWithLogFunction<Events::log> EventMark;
+typedef EventMarkWithLogFunction<Events::log_vm_operation> EventMarkVMOperation;
```
Comment (coleenp) [inline src/hotspot/share/utilities/events.hpp:471]: Can you add a comment just above EventMarkVMOperation line 470 like,
// These go to the vm_operation log
For some reason, I read the comment as applying to both instantiations.

**Review by coleenp — APPROVED** [review]
This looks good. Took a while to figure out why LogFunction was added, but this is helpful.  We might want to separate class loading in the future.

File: src/hotspot/share/utilities/events.hpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/utilities/events.hpp:473]: Thanks for reviewing.

**Review by dholmes-ora — APPROVED** [review]
Hi Stefan,

Adding the new event buffers itself seems fine.

Unlike Coleen I can't figure out why you added LogFunction and the new template classes.  :) I can't tell if this was necessary or you just preferred to re-do the specialization mechanism. ??

Thanks,
David

**Comment by stefank** [comment]
> Hi Stefan,
> 
> Adding the new event buffer itself seems fine.
> 
> Unlike Coleen I can't figure out why you added LogFunction and the new template classes. :) I can't tell if this was necessary or you just preferred to re-do the specialization mechanism. ??
> 
> Thanks,
> David

It was mostly done to not duplicate the code of EventMark when adding new EventMarkX classes. This can probably be done different ways. I'm open to suggestions.

File: src/hotspot/share/utilities/events.hpp [hunk 2]
```diff
@@ -414,16 +428,48 @@ inline void EventLogBase<ExtendedStringLogMessage>::print(outputStream* out, Ext
...
+  EventLogFunction _log_function;
+  StringLogMessage _buffer;
+
+ protected:
+  void log_start(const char* format, va_list argp) ATTRIBUTE_PRINTF(2, 0);
+  void log_end();
+
+ public:
+  EventMarkBase(EventLogFunction log_function);
+};
```
Comment (tschatzl) [inline src/hotspot/share/utilities/events.hpp:443]: The `EventMarkBase` constructor could be protected as well, and the class `NONCOPYABLE`.

**Review by tschatzl — APPROVED** [review]
Lgtm, feel free to ignore the comment on the protected constructor/NONCOPYABLE.

File: src/hotspot/share/utilities/events.hpp (hunk 2 above)
Comment (stefank) [reply src/hotspot/share/utilities/events.hpp:443]: Will update. Thanks for reviewing.

**Comment by coleenp** [comment]
I think the way Stefan did it would make it easier to add Class Loading specific EventMark.
classfile/classLoader.cpp:  EventMark m("loading class %s", class_name);

Although I don't see any other EventMarks except for shenandoah GC.

So there could be future refactoring here.

**Comment by stefank** [comment]
> > Hi Stefan,
> > Adding the new event buffer itself seems fine.
> > Unlike Coleen I can't figure out why you added LogFunction and the new template classes. :) I can't tell if this was necessary or you just preferred to re-do the specialization mechanism. ??
> > Thanks,
> > David
> 
> It was mostly done to not duplicate the code of EventMark when adding new EventMarkX classes. This can probably be done different ways. I'm open to suggestions.

@dholmes-ora Have you had time to think about this?

**Comment by stefank** [comment]
Thanks for the reviews!
/integrate

**Pull request closed without merging**